Move device selection of adc driver into driver's own kconfig

cybertale opened the pull request:
The selection of which device and channel to enable was configured
through Kconfig under the 'adc' folder, and it offers two selections:
ADC_0 or ADC_1. The problem is:
1. Some drivers only have one ADC instance and some have more than two.
2. Some socs like stm32 numbering its ADC starting from 1, not 0, so the
first device for stm32 should be ADC_1.
This commit moves the selection of device into every driver's own
kconfig file, so it can be customized.
Also, remove CONFIG_ADC_0 from prj_base.conf of adc_api test since every
driver may have different kconfig options to open their own first
device.
Two rules are applied to kconfig files:
1. If a soc has only one ADC instance, there's no need to manually
enable this instance.
2. If a soc has more than one ADC instance, the first ADC instance is
selected default. So there is always at lease one instance available
when the driver is enabled.

anangl commented on drivers/adc/Kconfig.intel_quark:
```diff
@@ -9,12 +9,15 @@
 
+config ADC_0
+	bool "Enable ADC 0"
+	default y
```
I'm wondering, will there be a use case for not enabling the only ADC instance while enabling the ADC driver? Maybe the instance selection should be just dropped for the drivers that handle only one ADC instance?

cybertale replied:
Both ways are good for me. I was thinking that someone may enable the ADC driver and forget to enable ADC instances because he thinks there is only one ADC instance. This way no one forgets to enable ADC instance.
Removing this forces people to select instances, so they would have a clear mind about which instances they enabled.

anangl replied:
> Removing this forces people to select instances, so they would have a clear mind about which instances they enabled.

I am not sure if we understand each other well. What I was trying to propose is not to just remove `default y`, I was thinking about removing the whole `ADC_0` option. My point is that for drivers that handle only one instance this option does not seem to make much sense. It will probably always need to be enabled (either by `default y` or explicitly) anyway. What would be the point of enabling the ADC driver and not enabling any ADC instances?

cybertale replied:
Yes, I agree there's no point to select a driver without selecting an instance. I was thinking about the consistency, if someone has been using ADC with one instance all the time, and the time he's going to start to use a multiple instances ADC. he might forget to choose any instances.
So, how about this, remove 'ADC_0' from every single instance drivers, and add 'default y' to ADC_0 to the ones with multiple instances.

anangl replied:
> So, how about this, remove 'ADC_0' from every single instance drivers, and add 'default y' to ADC_0 to the ones with multiple instances.

Yes, that's exactly what I was thinking about. Such approach should be convenient for everyone - no need for this a bit artificial `#ifdef CONFIG_ADC_0` in single instance drivers, and in every case some ADC instance enabled, so any sample or test application that just sets `CONFIG_ADC=y` has some device to work with.

cybertale committed:
kconfig: drivers: adc: move instances' selection into driver's kconfig

The selection of which device and channel to enable was configured
through Kconfig under the 'adc' folder, and it offers two selections:
ADC_0 or ADC_1. The problem is:
1. Some drivers only have one ADC instance and some have more than two.
2. Some socs like stm32 numbering its ADC starting from 1, not 0, so the
first device for stm32 should be ADC_1.
This commit moves the selection of device into every driver's own
kconfig file, so it can be customized.
Also, remove CONFIG_ADC_0 from prj_base.conf of adc_api test since every
driver may have different kconfig options to open their own first
device.
Two rules are applied to kconfig files:
1. If a soc has only one ADC instance, there's no need to manually
enable this instance.
2. If a soc has more than one ADC instance, the first ADC instance is
selected default. So there is always at lease one instance available
when the driver is enabled.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

anangl reviewed (approved):
I think you still need to remove this tab character from `drivers/adc/Kconfig.sam_afec:21`, as checkpatch will keep complaining about it.

ioannisg commented:
@cybertale do you want to get this into 2.0 release? If so, please rebase the PR.

MaureenHelm reviewed (changes requested):
No, I don't think we should go in this direction. We would then have to create driver-specific instance configs for spi, i2c, uart, etc. Also, some drivers can have a different number of instances depending on the soc. I think instead we should be doing this in dts somehow.

cybertale commented:
I also think this is not the best approach now. Which device to be turned on should be configured in the board configuration file. Thanks for your time, PR closed.
